lib/hw_unique_key/hw_unique_key.c

```diff
@@ -5,57 +5,226 @@
 					      sizeof(huk_magic));
 	if (err != 0) {
-		printk("The HUK loading failed "
-		       "with error code: %d\n\r",
-		       err);
-		NVIC_SystemReset();
+		printk("The HUK loading failed with error code: %d\n\r", err);
+		k_panic();
 	}
 
 	/* Lock the flash page which holds the key */
 	err = fprotect_area_no_access(huk_addr, CONFIG_FPROTECT_BLOCK_SIZE);
 	if (err != 0) {
 		printk("Fprotect failed with error code: %d\n\r", err);
-		NVIC_SystemReset();
+		k_panic();
 	}
 
 	printk("The Hardware Unique Key loaded successfully!\n\r");
-	return 0;
 }
+
+#else
+
+#define KMU_KEYSLOT_SIZE_WORDS 4
+
+/* Check whether a Hardware Unique Key has been written to the KMU. */
+static bool key_written(uint32_t kmu_slot)
+{
+	uint32_t idx = UICR_KEY_INDEX(kmu_slot);
+
+	NRF_KMU->SELECTKEYSLOT = kmu_slot;
+	if (NRF_UICR_S->KEYSLOT.CONFIG[idx].PERM != 0xFFFFFFFF) {
+		NRF_KMU->SELECTKEYSLOT = 0;
+		return true;
+	}
+	if (NRF_UICR_S->KEYSLOT.CONFIG[idx].DEST != 0xFFFFFFFF) {
+		NRF_KMU->SELECTKEYSLOT = 0;
+		return true;
+	}
+	for (int i = 0; i < KMU_KEYSLOT_SIZE_WORDS; i++) {
+		if (NRF_UICR_S->KEYSLOT.KEY[idx].VALUE[i] != 0xFFFFFFFF) {
+			NRF_KMU->SELECTKEYSLOT = 0;
+			return true;
+		}
+	}
+	NRF_KMU->SELECTKEYSLOT = 0;
+	return false;
+}
+
+
+#if HUK_HAS_CC310
+void hw_unique_key_write(uint32_t kmu_slot, const uint8_t *key)
+{
+	int err = -1;
+
+	if (kmu_slot == HUK_KEYSLOT_KDR) {
+		err = nrf_cc3xx_platform_kmu_write_kdr_slot(key);
+	} else {
+		err = nrf_cc3xx_platform_kmu_write_key_slot(
+			UICR_KEY_INDEX(kmu_slot), NRF_CC3XX_PLATFORM_KMU_AES_ADDR,
+			0xFFFF0000 | UICR_KEYSLOT_CONFIG_PERM_PUSH_Msk, key);
+	}
+
+	if (err != 0) {
+		printk("The HUK writing (%d) failed with error code: %d\n\r",
+			err, kmu_slot);
+		k_panic();
+	}
+}
+
+bool hw_unique_key_is_written(uint32_t kmu_slot)
+{
+	return key_written(kmu_slot);
+}
+
+void hw_unique_key_load_kdr(void)
+{
+	int err = nrf_cc3xx_platform_kmu_push_kdr_slot_and_lock();
+
+	if (err != 0) {
+		printk("The HUK loading failed with error code: %d\n\r", err);
+		k_panic();
+	}
+}
+
+
+#elif HUK_HAS_CC312
+void hw_unique_key_write(uint32_t kmu_slot, const uint8_t *key)
+{
+	int err = nrf_cc3xx_platform_kmu_write_key_slot(
+			UICR_KEY_INDEX(kmu_slot), NRF_CC3XX_PLATFORM_KMU_AES_ADDR,
+			0xFFFF0000 | UICR_KEYSLOT_CONFIG_PERM_PUSH_Msk, key);
+	if (err == 0) {
+		err = nrf_cc3xx_platform_kmu_write_key_slot(
+				UICR_KEY_INDEX(kmu_slot) + 1, NRF_CC3XX_PLATFORM_KMU_AES_ADDR_2,
+				0xFFFF0000 | UICR_KEYSLOT_CONFIG_PERM_PUSH_Msk,
+				key + (HUK_SIZE_BYTES / 2));
+	}
+
+	if (err != 0) {
+		printk("The HUK writing (%d) failed with error code: %d\n\r",
+			err, kmu_slot);
+		k_panic();
+	}
+}
+
+bool hw_unique_key_is_written(uint32_t kmu_slot)
+{
+	return key_written(kmu_slot) || key_written(kmu_slot + 1);
+}
+
+#endif /* HUK_HAS_CC310 */
+#endif /* HUK_HAS_KMU */
+
+#ifdef CONFIG_HW_UNIQUE_KEY_RANDOM
+void hw_unique_key_write_random(void)
+{
+	uint8_t rand[HUK_SIZE_BYTES * sizeof(slots)];
```

Comment:
I would consider a different name than rand, as rand is a C library function to generate random numbers

Reply:
Changed to `rand_bytes`